File: gunhoflash/search-in-rotated-sorted-array.py
from typing import List

class Solution:
  def search(self, nums: List[int], target: int) -> int:
    leftIndex = 0
    rightIndex = len(nums) - 1

    while leftIndex < rightIndex:
      midIndex = (leftIndex + rightIndex) // 2
      leftValue = nums[leftIndex]
      midValue = nums[midIndex]
      rightValue = nums[rightIndex]

      if midValue == target: return midIndex

      if leftValue <= midValue:
        if leftValue <= target and target <= midValue:
          rightIndex = midIndex
        else:
          leftIndex = midIndex + 1
      elif midValue <= target and target <= rightValue:
        leftIndex = midIndex + 1
      else:
        rightIndex = midIndex

    return leftIndex if nums[leftIndex] == target else -1

File: gunhoflash/search-in-rotated-sorted-array.ts
function search(nums: number[], target: number): number {
  let leftIndex = 0;
  let rightIndex = nums.length - 1;

  while (leftIndex < rightIndex) {
    const midIndex = Math.floor((leftIndex + rightIndex) / 2);
    const leftValue = nums[leftIndex];
    const midValue = nums[midIndex];
    const rightValue = nums[rightIndex];

    if (midValue === target) return midIndex;

    if (leftValue <= midValue) {
      if (leftValue <= target && target <= midValue) {
        rightIndex = midIndex;
      } else {
        leftIndex = midIndex + 1;
      }
    } else if (midValue <= target && target <= rightValue) {
      leftIndex = midIndex + 1;
    } else {
      rightIndex = midIndex;
    }
  }

  return nums[leftIndex] === target ? leftIndex : -1;
}
